editoast/simulation/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+    type Context: Send;
+
+    /// Number of cache entry read attempts per batch
+    ///
+    /// Choose this value based on the size of simulation results and acceptable latency considering all reads hit.
+    const CACHE_READS_BATCH_SIZE: usize;
+
+    /// Computes the cache key based on simulation inputs
+    fn key(&self, app_version: &str) -> String;
+
+    /// Performs the simulation according to inputs and context
+    ///
+    /// This function does **not** need to handle any caching concerns.
+    fn compute(
+        self,
+        ctx: Self::Context,
+    ) -> impl Future<Output = Result<Self::Output, Self::Error>> + Send;
+
+    /// Retrieves the simulation result from cache or performs the simulation
+    ///
+    /// To batch simulations, check out [trait SimulationStreamExt].
+    ///
+    /// # Errors
+    ///
+    /// Should any caching error occur while reading, the simulation is performed.
+    /// Cache write errors are ignored. So are serde errors for caching.
+    /// All errors are logged.
+    #[expect(async_fn_in_trait)] // not for public (ie. outside editoast) use, no auto traits bounds to specify on the resulting future
+    async fn simulate(
+        self,
+        mut vkconn: cache::Connection,
+        ctx: Self::Context,
+    ) -> Result<Self::Output, Self::Error> {
+        let key = self.key(vkconn.app_version());
+        match vkconn
+            .json_get::<Self::Output, _>(&key)
+            .await
+            .unwrap_or_else(|e| {
+                tracing::error!(?e, key, "cache read error — performing simulation");
+                None
+            }) {
+            Some(value) => Ok(value),
+            None => {
+                let value = self.compute(ctx).await?;
+                match serde_json::to_string(&value) {
+                    Err(e) => {
+                        tracing::error!(
+                            ?e,
+                            key,
+                            "serialization error before cache write — skipping cache write"
+                        );
+                    }
+                    Ok(serialized) => {
+                        tokio::spawn(async move {
+                            use deadpool_redis::redis::AsyncCommands as _;
+                            if let Err(e) = vkconn.set::<_, _, ()>(key, serialized).await {
+                                tracing::error!(?e, "cache write error");
+                            }
+                        });
+                    }
+                };
+                Ok(value)
+            }
+        }
+    }
+}
+
+/// A named tuple for a value with a correlation key
+pub struct Correlated<CorrelationKey, T> {
+    pub correlation_key: CorrelationKey,
+    pub data: T,
+}
+
+impl<CorrelationKey, T> Correlated<CorrelationKey, T> {
+    pub fn into_parts(self) -> (CorrelationKey, T) {
+        (self.correlation_key, self.data)
+    }
+}
```

Comment:
Why not using `From` trait?
```suggestion
impl<CorrelationKey, T> From<Correlated<CorrelationKey, T>> for (CorrelationKey, T) {
    fn from(correlated: Correlated<CorrelationKey, T>) -> Self {
        (correlated.correlation_key, correlated.data)
    }
}
```

Reply:
Done in 82464681e, I find it harder to understand the conversion, fine with both, lmk what you prefer